atef/report.py

```diff
@@ -0,0 +1,546 @@
+"""
+Report rendering framework
+"""
+
+import hashlib
+from dataclasses import fields
+from datetime import datetime
+from pathlib import Path
+from typing import Any, Dict, Generator, List, Optional, Tuple, Union
+
+from reportlab import platypus
+from reportlab.lib import colors, enums, pagesizes, units
+from reportlab.lib.styles import ParagraphStyle as PS
+from reportlab.lib.styles import getSampleStyleSheet
+from reportlab.lib.units import cm
+from reportlab.pdfgen.canvas import Canvas
+from reportlab.platypus.doctemplate import BaseDocTemplate, PageTemplate
+from reportlab.platypus.flowables import Flowable
+from reportlab.platypus.frames import Frame
+from reportlab.platypus.paragraph import Paragraph
+from reportlab.platypus.tableofcontents import TableOfContents
+
+from atef.check import Result
+from atef.config import (PreparedComparison, PreparedConfiguration,
+                         PreparedDeviceConfiguration, PreparedFile,
+                         PreparedGroup, PreparedPVConfiguration,
+                         PreparedSignalComparison, PreparedToolComparison,
+                         PreparedToolConfiguration)
+from atef.enums import Severity
+
+h1 = PS(name='Heading1', fontSize=16, leading=20)
+
+h2 = PS(name='Heading2', fontSize=14, leading=17)
+
+l0 = PS(name='list0', fontSize=12, leading=15, leftIndent=0,
+        rightIndent=0, spaceBefore=12, spaceAfter=0)
+
+styles = getSampleStyleSheet()
+
+
+class AtefReport(BaseDocTemplate):
```

Comment:
The report is pretty awesome!

We have some tweaks to do, of course, but I'm not going to waste either of our time nitpicking on reportlab stuff. Its API doesn't make anything easy...

Reply:
If you do have thoughts for tweaks/nitpicks, let me know and I can make an issue.  I know I'll forget otherwise